File: assemblyadapter-core/src/main/java/com/github/panpf/assemblyadapter/AssemblyItem.java
package com.github.panpf.assemblyadapter;

import android.content.Context;
import android.content.res.Resources;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public abstract class AssemblyItem<DATA> implements Item<DATA> {

    @NonNull
    public final Context context;
    @NonNull
    public final Resources resources;
    @NonNull
    public final View itemView;

    @Nullable
    private DATA data;
    private int position = -1;

    public AssemblyItem(@NonNull View itemView) {
        this.itemView = itemView;
        this.context = itemView.getContext();
        this.resources = itemView.getResources();
    }

    public AssemblyItem(int itemLayoutId, @NonNull ViewGroup parent) {
        this(LayoutInflater.from(parent.getContext()).inflate(itemLayoutId, parent, false));
    }

    @Override
    public void dispatchBindData(int position, @Nullable DATA data) {
        this.position = position;
        this.data = data;
        bindData(position, data);
    }

    protected abstract void bindData(int position, @Nullable DATA data);

    @NonNull
    @Override
    public final View getItemView() {
        return this.itemView;
    }

    @Nullable
    @Override
    public DATA getData() {
        return data;
    }

    @Override
    public int getPosition() {
        return position;
    }
}

File: assemblyadapter-common-item/src/main/java/com/github/panpf/assemblyadapter/AssemblyItem.kt
package com.github.panpf.assemblyadapter

import android.content.Context
import android.content.res.Resources
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

// todo 直接命名为 Item
abstract class AssemblyItem<DATA>(val itemView: View) {

    private var _data: DATA? = null
    private var _bindingAdapterPosition: Int = -1
    private var _absoluteAdapterPosition: Int = -1

    val context: Context = itemView.context
    val appContext: Context = context.applicationContext
    val resources: Resources = context.resources

    val dataOrNull: DATA?
        get() = _data
    val requireData: DATA
        get() = _data!!
    val bindingAdapterPosition: Int
        get() = _bindingAdapterPosition
    val absoluteAdapterPosition: Int
        get() = _absoluteAdapterPosition

    constructor(itemLayoutId: Int, parent: ViewGroup) : this(
        LayoutInflater.from(parent.context).inflate(itemLayoutId, parent, false)
    )

    open fun dispatchBindData(bindingAdapterPosition: Int, absoluteAdapterPosition: Int, data: DATA) {
        this._data = data
        this._bindingAdapterPosition = bindingAdapterPosition
        this._absoluteAdapterPosition = absoluteAdapterPosition
        bindData(_absoluteAdapterPosition, data)
    }

    protected abstract fun bindData(bindingAdapterPosition: Int, data: DATA)
}